src/Vulkan/Silk.NET.Vulkan/Vk.cs

```diff
@@ -154,34 +157,47 @@ protected override void PostInit()
         /// Checks whether the given instance extension is available.
         /// </summary>
         /// <param name="extension">The instance extension name.</param>
+        /// <param name="layer">The instance layer name.</param>
         /// <returns>Whether the instance extension is available.</returns>
-        public unsafe bool IsInstanceExtensionPresent(string extension)
+        public unsafe bool IsInstanceExtensionPresent(string extension, string layer = null)
         {
             // Note we use optimistic code to avoid locks, if this is called on multiple threads
             // then multiple initialisations can happen, only one thread will succeed in initialising the cache though.
             var cachedInstanceExtensions = _cachedInstanceExtensions;
             if (cachedInstanceExtensions is null)
             {
-                // Get count of properties
-                var instanceExtPropertiesCount = 0u;
-                EnumerateInstanceExtensionProperties((byte*) 0, &instanceExtPropertiesCount, null);
+                var layerPtr = layer == null ? 0 : SilkMarshal.StringToPtr(layer);
 
-                // Initialise return structure
-                using var mem = GlobalMemory.Allocate((int) instanceExtPropertiesCount * sizeof(ExtensionProperties));
-                var props = (ExtensionProperties*) Unsafe.AsPointer(ref mem.GetPinnableReference());
+                try {
+                    // Get count of properties
+                    var instanceExtPropertiesCount = 0u;
+                    EnumerateInstanceExtensionProperties((byte*) layerPtr, &instanceExtPropertiesCount, null);
+
+                    // Initialise return structure
+                    var propsArr = new ExtensionProperties[(int)instanceExtPropertiesCount];
 
-                // Get properties
-                EnumerateInstanceExtensionProperties((byte*) 0, &instanceExtPropertiesCount, props);
+                    fixed(ExtensionProperties* props = propsArr) {
+                        // Get properties
+                        EnumerateInstanceExtensionProperties((byte*) layerPtr, &instanceExtPropertiesCount, props);
+
+                        cachedInstanceExtensions = new HashSet<string>();
+                        for (var p = 0; p < instanceExtPropertiesCount; p++)
+                        {
+                            cachedInstanceExtensions.Add(Marshal.PtrToStringAnsi((nint) props[p].ExtensionName));
```

Comment:
It is not completely out of the realm of possibility that the same extension may have multiple hits. For example, consider the official Khronos Extension Layer which provides polyfill implementations where the driver does not natively support an extension, if this is being used alongside a driver that actually does support it (e.g. one polyfill is being used from the layer that isn't implemented, but the layer also provides other extensions which are implemented) this extension name would be returned for both `null` and `"VK_LAYER_KHRONOS_shader_object"` for example.